src/neorados/cls/lock.h

```diff
@@ -0,0 +1,651 @@
+  /// \param token Boost.Asio completion token
+  ///
+  /// \return Possibly errors in a way appropriate to the completion
+  /// token. See Boost.Asio documentation.
+  template<boost::asio::completion_token_for<
+	     void(boost::system::error_code)> CompletionToken>
+  auto break_lock(std::string other_cookie, entity_name_t locker,
+		  CompletionToken&& token) {
+    return r.execute(obj, ioc,
+		     lock::break_lock(WriteOp{}, name, std::move(other_cookie),
+				      std::move(locker)),
+		     std::forward<CompletionToken>(token));
+  }
+
+  /// \brief Get info for a named lock
+  ///
+  /// Execute an asynchronous operation that returns all information on
+  /// the named lock
+  ///
+  /// \param token Boost.Asio CompletionToken
+  ///
+  /// \return The lock list in a way appropriate to the completion
+  /// token. See Boost.Asio documentation.
+  template <boost::asio::completion_token_for<void(
+      boost::system::error_code, std::map<locker_id_t, locker_info_t>, type,
+      std::string)> CompletionToken>
+  auto info(CompletionToken&& token) {
+    return lock::info(r, obj, ioc, name,
+		      std::forward<CompletionToken>(token));
+  }
+
+  /// \brief Change the lock's cookie
+  ///
+  /// Execute an asynchronous operation that sets the lock's cookie to
+  /// a new value.
+  ///
+  /// \param t Must match type of existing lock
+  /// \param token Boost.Asio CompletionToken
+  /// \param new_cookie The new cookie
+  ///
+  /// \return Errors as appropriate to the completion token. See Boost.Asio
+  /// documentation.
+  template<typename CompletionToken>
+  auto set_cookie(type t, std::string new_cookie,
+		  CompletionToken &&token) {
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmismatched-new-delete"
+    namespace asio = boost::asio;
+    using boost::system::error_code;
+    using boost::system::system_error;
+    using boost::system::errc::no_message_available;
+
+  return asio::async_initiate<CompletionToken,
+			      void(error_code)>
+    (asio::experimental::co_composed<void(error_code)>
+     ([](auto state, RADOS& r, Object oid, IOContext ioc,
+	 std::string name, type t, std::string& cookie,
+	 std::string tag, std::string new_cookie) -> void {
+       try {
+	 co_await r.execute(oid, ioc,
+			    lock::set_cookie(WriteOp{}, std::move(name), t,
+					     cookie, std::move(tag), new_cookie),
+			    asio::deferred);
+	 cookie.swap(new_cookie);
+       } catch (const system_error& e) {
+	 co_return e.code();
+       }
+       co_return error_code{};
+     }, r.get_executor()),
+     token, std::ref(r), obj, ioc, name, t, std::ref(cookie), tag,
+     std::move(new_cookie));
+#pragma GCC diagnostic pop
+  }
+
+  /// \brief Return a randomly cookie
+  ///
+  /// \return A randomly generated cookie
+  std::string random_cookie() {
+    static constexpr std::size_t COOKIE_LEN = 16;
+    std::string s(COOKIE_LEN, '\0');
+    gen_rand_alphanumeric(r.cct(), s.data(), COOKIE_LEN);
+    return s;
```

Comment:
this overload of `gen_rand_alphanumeric()` would only give you 15 characters plus a null. there's a more convenient overload that returns `std::string` directly:
```c++
return gen_rand_alphanumeric(r.cct(), COOKIE_LEN);
```